docs/presentation/Hierarchies/GroupingRule.md

```diff
@@ -7,99 +7,282 @@ Grouping rules provide advanced ways to group instances when creating hierarchie
+| `imageId`   | No        | `string` | [`imageId` of the property group specification](#attribute-imageid) | Identifier of an image to use for the grouping node.                    |
+| `label`     | No        | `string` | `"{from value} - {to value}"`                                       | Grouping node's label. May be [localized](../Advanced/Localization.md). |
+
+Range [`fromValue`, `toValue`] is inclusive on both sides. If a value falls into more than one range, the first listed range that contains the value is chosen.
+
+```ts
+[[include:Hierarchies.Grouping.PropertyGroup.Ranges.Ruleset]]
+```
+
+![Example of using "ranges" attribute](./media/hierarchy-grouping-with-ranges-attribute.png)
+
+#### Deprecated Attribute: `groupingValue`
+
+> **Default value:** `"DisplayLabel"`
+
+Specifies whether instances should be grouped using property's display or raw value.
+
+> **Note:** Grouping by property value is required if the display label is overridden to display grouped instances count.
+>
+> **Warning:** Grouping by label and sorting by property value is not possible.
+
+Display value should always be used for grouping. In cases when there's a need to show grouped instances count suffix, that
+can be achieved at the UI component layer by composing UI node's label from node's display label and [GroupingNodeKey.groupedInstancesCount]($presentation-common).
+
+#### Deprecated Attribute: `sortingValue`
+
+> **Default value:** `"DisplayLabel"`
+
+Specifies whether nodes should be sorted by their display label or the grouping property's value. In most cases the result is the same,
+unless a [label override rule](../Customization/LabelOverride.md) is used to change node's display label.
+
+> **Note:** Sorting by property value only makes sense when instances are grouped by property value as well.
+>
+> **Warning:** Grouping by label and sorting by property value is not possible.
+
+### Same Label Instance Grouping
+
+Allows grouping multiple instances with the same label into one ECInstances type of node. Similar to display label grouping, but instead of showing a grouping
+node with multiple grouped ECInstance nodes, it shows a single ECInstances node which represents multiple ECInstances.
+
+| Name                                              | Required? | Type                       | Default   |
+| ------------------------------------------------- | --------- | -------------------------- | --------- |
+| [`applicationStage`](#attribute-applicationstage) | No        | `"Query" \| "PostProcess"` | `"Query"` |
+
+#### Attribute: `applicationStage`
+
+> **Default value:** `"Query"`
+
+> **Performance note:** Grouping at post-processing step requires loading the whole hierarchy level before
+> returning even the first node - avoid using with large numbers of nodes.
+
+Stage of hierarchy creation at which the rule is applied.
```

Comment:
Restructured description to localise all the important information into one place. Left the code samples untouched. Two samples is perhaps too much, maybe display a sample ruleset that `"Query"` is unable to handle and show how `"PostProcessing"` makes grouping work correctly? Then exchange test code snippets with tree screenshots.

```suggestion
#### Attribute: `applicationStage`

> **Default value:** `"Query"`

Grouping nodes by label is an expensive operation because it requires the whole hierarchy level to be created before even the first grouped node can be produced. To alleviate the performance impact when this specification is used, two `applicationStage` settings have been introduced:

- `"Query"` groups instances during ECSql query, which can often make use of database indices and is generally fairly quick. It is chosen as the default option, however, it fails to produce grouping nodes when certain ruleset specifications are involved.
- `"PostProcess"` groups instances after the whole hierarchy level is built. It incurs a large performance penalty, but it will produce the expected result in all cases.

#### Choosing between `"Query"` and `"PostProcess"`

Always prefer `"Query"` unless any of the following conditions apply to the hierarchy level that is being grouped:

- The hierarchy level is built out of instances that do not share a common base class.
- The hierarchy level aggregates nodes that have been produced by more than one specification.
- The hierarchy level contains nodes that have been merged from a nested hierarchy level as a result of [`hideNodesInHierarchy`](./InstanceNodesOfSpecificClasses.md#attribute-hidenodesinhierarchy) or [`hideExpression`](./InstanceNodesOfSpecificClasses.md#attribute-hideexpression) attributes.

On the other hand, `"PostProcess"` can be applied to any hierarchy level, regardless of its composition, but at a cost to performance.
```

Reply:
Applied your suggestion and updated the example - please have a look